### ISSUE TITLE
Undefined values of variables destructed in generator declaration

### ISSUE THREAD

**Issue by ceresnam** [microsoft/TypeScript#30746]
**TypeScript Version:**  3.4.1  (works in 3.2.4)

**Search Terms:** 
destruct  generator

**Code**

```ts
export function* saveLike({ entity, id, xhr }: LikeToggleAction) {

    if (!(yield call(checkIfUserIsLoggedIn, ACTION_TO_ADD_REPLY))) {
        return;
    }

    // at this point entity, id, xhr are undefined

    // ... other code
}
```

**Expected behavior:**
Values of variables destructed from function declaration should not be undefined, i.e. should have values that were passed to function call

**Actual behavior:**
After first resume of generator, destructed values are lost (i.e. undefined). 

Problem is in transpiled code. Destructed code is inside case 0, i.e. subsequent calls to generator (case 1, case 2, ...) do not see the values.

```
function saveLike(_a) {
    var _b, key, current, newState, fakeResponse, error_1;    
    return tslib__WEBPACK_IMPORTED_MODULE_0__["__generator"](this, function (_c) {
        switch (_c.label) {
            case 0: 
                 var entity = _a.entity, id = _a.id, xhr = _a.xhr;
                 return [4 /*yield*/,  ...];
            case 1:
               // at this point entity, id, xhr are undefined
               // ...
```

**Comment by pmartin-veritone** [comment]
This issue forces redux-saga users to stay with v3.3.

**Comment by CherishIt** [comment]
same problem

**Comment by ericanderson** [comment]
@RyanCavanaugh Just hit us too. Repro:


```ts
function* doStuff(obj, { becomesUndefined }) {
    const styles = { [obj.id]: obj };

    yield 1;
    alert(becomesUndefined);
}

const q = doStuff("", { becomesUndefined: 5 });
q.next();
q.next();
```

https://www.typescriptlang.org/play/#src=function*%20doStuff(obj%2C%20%7B%20becomesUndefined%20%7D)%20%7B%0D%0A%20%20%20%20const%20styles%20%3D%20%7B%20%5Bobj.id%5D%3A%20obj%20%7D%3B%0D%0A%0D%0A%20%20%20%20yield%201%3B%0D%0A%20%20%20%20alert(becomesUndefined)%3B%0D%0A%7D%0D%0A%0D%0Aconst%20q%20%3D%20doStuff(%22%22%2C%20%7B%20becomesUndefined%3A%205%20%7D)%3B%0D%0Aq.next()%3B%0D%0Aq.next()%3B

**Comment by rbuckton** [comment]
This was fixed by #30726.
